cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,367 @@
+        stochastic_silencing_rate: Rate at which to randomly drop out cassettes,
+            to simulate dropout due to low sensitivity of assays.
+        heritable_missing_data_state: Integer representing data that has gone
+            missing due to a heritable event (i.e. Cas9 resection or heritable
+            silencing).
+        stochastic_missing_data_state: Integer representing data that has
+            gone missing due to the stochastic dropout from single-cell assay
+            sensitivity.
+        random_seed: Numpy random seed to use for deterministic simulations.
+            Note that the numpy random seed gets set during every call to
+            `overlay_data`, thereby producing deterministic simulations every
+            time this function is called.
+
+    Raises:
+        DataSimulatorError if assumptions about the system are broken.
+    """
+
+    def __init__(
+        self,
+        number_of_cassettes: int = 10,
+        size_of_cassette: int = 3,
+        mutation_rate: Union[float, List[float]] = 0.01,
+        state_generating_distribution: Callable[[], float] = lambda: np.random.exponential(
+            1e-5
+        ),
+        number_of_states: int = 100,
+        state_priors: Optional[Dict[int, float]] = None,
+        heritable_silencing_rate: float = 1e-4,
+        stochastic_silencing_rate: float = 1e-2,
+        heritable_missing_data_state: int = -1,
+        stochastic_missing_data_state: int = -1,
+        random_seed: Optional[int] = None,
+    ):
+
+        if number_of_cassettes <= 0 or not isinstance(number_of_cassettes, int):
+            raise DataSimulatorError("Specify a positive number of cassettes.")
+        if size_of_cassette <= 0 or not isinstance(size_of_cassette, int):
+            raise DataSimulatorError(
+                "Specify a positive number of cut-sites" " per cassette."
+            )
+
+        self.size_of_cassette = size_of_cassette
+        self.number_of_cassettes = number_of_cassettes
+
+        if isinstance(mutation_rate, float):
+            if mutation_rate < 0:
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
+                )
+            number_of_characters = size_of_cassette * number_of_cassettes
+            self.mutation_rate_per_character = [
+                mutation_rate
+            ] * number_of_characters
+        else:
+            if len(mutation_rate) != (
+                self.number_of_cassettes * self.size_of_cassette
+            ):
+                raise DataSimulatorError(
+                    "Length of mutation rate array is not"
+                    " the same as the number of characters."
+                )
+
+            if np.any(np.array(mutation_rate) < 0):
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
+                )
+            self.mutation_rate_per_character = mutation_rate
```

Comment:
Let's make a copy to avoid exposing the internal representation of the class (similar to the issue we had with CassiopeiaTree). E.g.: `self.mutation_rate_per_character = mutation_rate[:]`

Reply:
Good call, this is done in my latest commit.